DOCSP-51698-llm-rag-claim-management (#282)
mongodb/docs-atlas-architecture: source/solutions-library/claim-management.txt

   :depth: 1
   :class: singlecol

> Discover how to combine Atlas Vector Search and large language models (LLMs) to 

Reviewer: Large Language Models

> (LLMs) to streamline the claim adjustment process.

Reviewer 2: Generally, the term is "claims adjustment/adjuster". We probably want to rename the title as well.
```suggestion
(LLMs) to streamline the claims adjustment process.
```

Solution Overview
-----------------

One of the biggest challenges for claims adjusters is aggregating information
from disparate systems and diverse data formats. Over the years, insurance
companies have accumulated terabytes of `unstructured data
<https://www.mongodb.com/unstructured-data>`__ in their datastores, which can
help uncover business insights, deliver better customer experiences, and
streamline operations. However, many companies fail to capitalize on this.

To benefit these organizations and their customers, MongoDB's solution combines
the power of `Atlas Vector Search
<https://www.mongodb.com/products/platform/atlas-vector-search>`__ and `Large
Language Models (LLMs) <https://www.mongodb.com/basics/large-language-models>`__
in a `retrieval augmented generation (RAG)
<https://www.mongodb.com/basics/retrieval-augmented-generation>`__ system, to

> allow organizations to go beyond the limitations of baseline foundational

Reviewer: and allows organization to go beyond ...

> models. Propriety data makes models context-aware, allowing them to leverage the

Reviewer: Use instead of leverage

Solution Overview
-----------------

> One of the biggest challenges for claim adjusters is aggregating information

Reviewer 2: ```suggestion
One of the biggest challenges for claims adjusters is aggregating information
```

from disparate systems and diverse data formats. Over the years, insurance
companies have accumulated terabytes of `unstructured data

> <https://www.mongodb.com/unstructured-data>`__ in their datastores, but fail to
> capitalize on it to uncover business insights, deliver better customer
> experiences, and streamline operations.

Reviewer 2: I think the negative sentiment of this sentence juxtaposed with the multiple benefits at the end reads a bit awkwardly. Would consider restructuring:
```suggestion
<https://www.mongodb.com/unstructured-data>`__ in their datastores, which can help
uncover business insights, deliver better customer experiences, and streamline operations.
However, many companies fail to capitalize on this.
```

To benefit these organizations and their customers, MongoDB's solution combines

> the power of `Altas Vector Search

Reviewer 2: ```suggestion
the power of `Atlas Vector Search
```

> <https://www.mongodb.com/products/platform/atlas-vector-search>`__ and a `Large

Reviewer 3: ```suggestion
<https://www.mongodb.com/products/platform/atlas-vector-search>`__ and `Large
```

Language Models (LLMs) <https://www.mongodb.com/basics/large-language-models>`__
in a `retrieval augmented generation (RAG)
<https://www.mongodb.com/basics/retrieval-augmented-generation>`__ system, to
allow organizations to go beyond the limitations of baseline foundational

> models. Propriety data makes models context aware, allowing them to leverage the

Reviewer 2: ```suggestion
models. Propriety data makes models context-aware, allowing them to leverage the
```

> leverage the full potential of AI to streamline operations.

Reviewer 3: remove passive voice

Reference Architecture
----------------------

With MongoDB

> ~~~~~~~~~~~~

Reviewer 2: i'd remove this, tbh

> To provide a unified development experience, documents are stored alongside

Reviewer 3: passive voice

their vector embeddings and associated metadata, eliminating the need to
retrieve data elsewhere. This allows users to focus on building their

> application, instead of maintaining a separate technology.

Reviewer 2: ```suggestion
application instead of maintaining a separate technology.
```

Ultimately, the data obtained from MongoDB Vector Search is fed to the LLM as 

> context.

Reviewer 2: might consider combining this sentence with the preceding paragraph so it's not standalone?

The process of the RAG querying flow is as follows:

#. The user writes a prompt in natural language.

#. Voyage AI's embedding model vectorizes the prompt.

#. Atlas Vector Search uses the vectorized prompt to retrieve relevant
   documents.

#. LLM uses both the context and original question to generate an answer.

> #. The user receives an answer.

Reviewer 2: awesome job adding these steps from the diagram!

      Create and configure an :ref:`Atlas Vector Search index
      <atlas-ui-create-vector-search>` for ``claimDescriptionEmbeddingCohere``
      called ``vector_index_claim_description_cohere``. You must structure the
      search index as follows:

      .. code-block:: json
         :copyable: true 

         {
            "fields": [
               { 
                  "type": "vector", 
                  "path": "claimDescriptionEmbeddingCohere",
                  "numDimensions": 350, 
                  "similarity": "cosine"
               }
            ]
         }

   .. step:: Configure the Backend
      
      Set up a virtual environment using Poetry.

   .. step:: Interact with the API

      Start the backend server. 

   .. step:: Connect to the Frontend

      Configure environment variables and run the frontend.

You have to run both the front and back end. You’ll access a web UI
that allows you to ask questions to the LLM, obtain an answer, and see the
reference documents used as context.

To try MongoDB's semantic search tool now, visit the :ref:`Atlas Vector Search
Quick Start guide <vector-search-quick-start>`.

Key Learnings

> -------------

Reviewer: For the key learning taking this approach:

- You can **create embeddings using different models and deployment options.** It is important to consider privacy and data protection requirements. You can deploy a model locally if your data needs to remain on the servers. Otherwise, you can call an API and get your vector embeddings back, as explained in [this](https://www.mongodb.com/docs/atlas/atlas-vector-search/create-embeddings/) tutorial. You can use [Voyage AI (acquired by MongoDB)](https://www.mongodb.com/blog/post/redefining-database-ai-why-mongodb-acquired-voyage-ai) or open-source models.

Data Model Approach
-------------------

In the demo solution, the data model is a simplified design that emulates
real-world insurance claim data. The approach leverages MongoDB's flexible
document model to handle the diverse data structure that stores embeddings
alongside their related document.

.. list-table::
   :header-rows: 1

>    :widths: 8 10 20

Reviewer 2: while i really appreciate finding different ways to display content (this is awesome!), we do want to try not to include large code blocks within tables because it kinda messes with SEO stuff. I might recommend breaking this table down :( sad!!!

          "_id": {
            "$oid": "65cc809c76da22d0089dfb2e"
          },
          "customerID": "c105",
          "policyNumber": "p105",
          "claimID": "cl105",
          "claimStatusCode": "Subrogation",
          "claimDescription": "High winds caused ...",
          "totalLossAmount": 4200,
          "claimFNOLDate": "2023-10-27",
          "claimClosedDate": "2024-09-01",
          "claimLineCode": "Auto",
          "damageDescription": "Roof caved in ...",
          "insurableObject": {
            "insurableObjectId": "abc105",
            "vehicleMake": "Make105",
            "vehicleModel": "Model105"
          },
          "coverages": [
            {
               "coverageCode": "888",
               "description": "3rd party responsible"
            },
            {
               "coverageCode": "777",
               "description": "Vehicle rental/loaner service for customer"
            }
          ],
          "claimDescriptionEmbedding": [-0.017, ..., 0.011],
          "damageDescriptionEmbedding": [-0.047, ..., -0.043],
          "photo": "105.jpg",
          "photoEmbedding": [9.629, ..., 14.075]
          }

Build the Solution
------------------

> For detailed setup instructions, follow the README of `this
> Github repository <https://github.com/mongodb-industry-solutions/RAG-Insurance>`__.

Reviewer 2: ```suggestion
For detailed setup instructions, follow the ``README`` of `this
GitHub repository <https://github.com/mongodb-industry-solutions/RAG-Insurance>`__.
```

You’ll be guided through the following steps:

To build the solution, follow these instructions:

.. procedure::
   :style: normal

   .. step:: Set up MongoDB database and collections

      Create a new database in MongoDB Atlas called ``demo_rag_insurance`` and
      use the provided dataset ``demo_rag_insurance_claims.json`` to create a
      collection called ``claims_final``.

   .. step:: Create a Vector Search Index

>       `Create and configure an Atlas Vector Search Index
>       <https://www.mongodb.com/docs/atlas/atlas-vector-search/vector-search-type/#create-an-atlas-vector-search-index>`__
>       for ``claimDescriptionEmbeddingCohere`` called

Reviewer 2: recommend replacing with `:ref:`:
```suggestion
      Create and configure an :ref:`Atlas Vector Search index <atlas-ui-create-vector-search>`
      for ``claimDescriptionEmbeddingCohere`` called
```

      ``vector_index_claim_description_cohere``. You must structure the search index as follows:

      .. code-block:: json
         :copyable: true 

         {
            "fields": [
               { 
                  "type": "vector", 
                  "path": "claimDescriptionEmbeddingCohere",
                  "numDimensions": 350, 
                  "similarity": "cosine"
               }
            ]
         }

   .. step:: Configure the Backend
      
      Set up a virtual environment using Poetry.

   .. step:: Interact with the API

      Start the backend server. 

   .. step:: Connect to the Frontend

      Configure environment variables and run the frontend.

> Ultimately, you have to run both the front and back end. You’ll access a web UI
> that allows you to ask questions to the LLM, obtain an answer, and see the
> reference documents used as context.

Reviewer 2: ```suggestion
You have to run both the front and back end. You’ll access a web UI
that allows you to ask questions to the LLM, obtain an answer, and see the
reference documents used as context.
```

> Visit the `Atlas Vector Search Quick Start guide
> <https://www.mongodb.com/docs/atlas/atlas-vector-search/tutorials/vector-search-quick-start/?tck=ai_as_web>`__
> to try our semantic search tool now.

Reviewer 2: Also ref here. And i think(?) we like to structure sentences as "to do A, visit B."
```suggestion
To try MongoDB's semantic search tool now, 
visit the :ref:`Atlas Vector Search Quick Start guide <vector-search-quick-start>`.
```

  create indexes for `local deployments

>   <https://www.mongodb.com/docs/atlas/cli/stable/atlas-cli-deploy-local/#std-label-atlas-cli-deploy-local-avs>`__.

Reviewer: What do you think if we reframe this section as this:

- You can **build Vector Search indexes** in MongoDB Atlas, alternatively, you can also build indexes for `local deployments <https://www.mongodb.com/docs/atlas/cli/stable/atlas-cli-deploy-local/#std-label-atlas-cli-deploy-local-avs>`__.

- **Performing a Vector Search query:** :ref:`Vector Search queries
  <return-vector-search-results>` have a dedicated operator within MongoDB’s
  :ref:`aggregation pipeline <aggregation>`. This means they can be concatenated
  with other operations, making it convenient for developers because they don’t

>   need to learn a different language or change context.

Reviewer: Similar here: 

- You can **run vector search queries with MongoDB's aggregation pipeline**, allowing you to concatenate multiple operations in your workflow. This approach eliminates the need to learn another programming language or change context.

- **Fast RAG Implementation:** Using the :ref:`LangChain <langchain>` framework
  that combines MongoDB Atlas Vector Search and the LLM allows for an easy and

>   fast RAG implementation.

Reviewer: Similar here:

- You can **develop a fast RAG implementation** with a LangChain framework that combines MongoDB Atlas Vector Search and LLMs.

> - **Performing a Vector Search query:** Notably, `Vector Search queries <https://www.mongodb.com/docs/atlas/atlas-vector-search/vector-search-stage/>`__ 
>   have a dedicated operator within MongoDB’s `aggregation pipeline

Reviewer 2: ```suggestion
- **Performing a Vector Search query:** :ref:`Vector Search queries <return-vector-search-results>`
  have a dedicated operator within MongoDB’s `aggregation pipeline
```

>   <https://www.mongodb.com/docs/manual/aggregation/>`__. This means they can be

Reviewer 2: ugh i cant comment on the above lines but replace general agg pipeline link with ref here:
```suggestion
  :ref:`aggregation pipeline <aggregation>`. This means they can be
```

  concatenated with other operations, making it convenient for developers
  because they don’t need to learn a different language or change context.

> - **Fast RAG Implementation:** Using the `LangChain
>   <https://www.mongodb.com/docs/atlas/ai-integrations/langchain/>`__ framework
>   that glues together MongoDB Atlas Vector Search and the LLM allows for an

Reviewer 2: suggestion: add ref and more plain language for readability
```suggestion
- **Fast RAG Implementation:** Using the :ref:`LangChain <langchain>` framework
  that combines MongoDB Atlas Vector Search and the LLM allows for an
```
